packages/block-library/src/table-of-contents/edit.js

```diff
@@ -145,6 +155,56 @@ export default function TableOfContentsEdit( {
 						}
 					/>
 				</ToolsPanelItem>
+				<ToolsPanelItem
+					hasValue={ () => ! includeAllHeadings }
+					label={ __( 'Include all headings' ) }
+					onDeselect={ () =>
+						setAttributes( { includeAllHeadings: true } )
+					}
+					isShownByDefault
+				>
+					<ToggleControl
+						__nextHasNoMarginBottom
+						label={ __( 'Include all headings' ) }
+						checked={ includeAllHeadings }
+						onChange={ ( value ) =>
+							setAttributes( { includeAllHeadings: value } )
+						}
+						help={
+							includeAllHeadings
+								? __(
+										'Including all heading levels in the table of contents.'
+								  )
+								: __(
+										'Filter headings by their level in the table of contents.'
+								  )
+						}
+					/>
+					{ ! includeAllHeadings && (
+						<SelectControl
+							__nextHasNoMarginBottom
+							__next40pxDefaultSize
+							label={ __( 'Include headings down to level' ) }
+							value={ maxLevel }
+							options={ [
+								{ value: 1, label: __( 'Heading 1' ) },
+								{ value: 2, label: __( 'Heading 2' ) },
+								{ value: 3, label: __( 'Heading 3' ) },
+								{ value: 4, label: __( 'Heading 4' ) },
+								{ value: 5, label: __( 'Heading 5' ) },
+								{ value: 6, label: __( 'Heading 6' ) },
+							] }
+							onChange={ ( value ) =>
+								setAttributes( {
+									maxLevel: parseInt( value, 10 ),
```

Comment:
FYI passing radix 10 isn't needed and hasn't been for like 10 years. `parseInt( value )` is enough